### PR TITLE
[WIP] Add meca content provider

### PR DESCRIPTION
**Pull request by stevejpurves**
This PR adds a content provider for MECA Bundles.

[MECA stands for Manuscript Exchange Common Approach](https://www.niso.org/standards-committees/meca), a NISO standard for transfer of manuscripts between manuscript systems.

The provider was written during in 2023 as part of the [AGU's NotebooksNow project](https://data.agu.org/notebooks-now/). The aim of that project is to make notebook based computational research papers a formal part of the scientific record, and has drawn on inputs from a number of working groups and stakeholder organisations. 

MECA was identified as a way to consistently package files from a git repository, paired with JATS xml, that would enable a set of notebooks, code, REES environment specification, into an single file that could be accepted and archived in existing publishing systems.

Support to produce MECA files has been included in Myst Markdown and Quatro author toolchains, enabling researchers to export to this format and the first AGU journal supporting the format is going live in 2024.

By adding provider as a standard provider in repo2docker/binderhub, we'll enable authors and researchers to reproduce research articles that include this bundle.

This provider is intended to be paired with the BinderHub Repo Provider being added in this PR:

### Operation

The provider receives a parsed url to a Meca bundle, which is a zip file. The zipfile is fetched to a local tmp folder and unpacked. Given a `manifest.xml` with expected tags to a `article-source-directory`, that identifies the path to use for docker creation. (we expect a REES compatible environment to be present. If this information is missing, we try to build from the root folder.

### CODE REVIEW

File: repo2docker/contentproviders/meca.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        for chunk in resp.iter_content(chunk_size=128):
+            dst.write(chunk)
+
+    return dst_filename
+
+
+def handle_items(_, item):
+    print(item)
+
+
```
Comment (sgaist) [inline repo2docker/contentproviders/meca.py:36]: This is unused
```suggestion
```

File: repo2docker/contentproviders/meca.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+def extract_validate_and_identify_bundle(zip_filename, dst_dir):
+    if not os.path.exists(zip_filename):
+        raise RuntimeError("Download MECA bundle not found")
```
Comment (sgaist) [inline repo2docker/contentproviders/meca.py:39]:
```suggestion
        raise RuntimeError("Downloaded MECA bundle not found")
```

File: repo2docker/contentproviders/meca.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.session.headers.update(
+            {
+                "user-agent": f"repo2docker MECA",
+            }
+        )
+
+    def detect(self, spec, ref=None, extra_args=None):
+        """`spec` contains a faux protocol of meca+http[s] for detection purposes
+        and we assume `spec` trusted as a reachable MECA bundle from an allowed origin
+        (binderhub RepoProvider class already checking for this).
```
Comment (sgaist) [inline repo2docker/contentproviders/meca.py:83]:
```suggestion
        (binderhub RepoProvider class is already checking for this).
```

File: repo2docker/contentproviders/meca.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        An other HEAD check in made here in order to get the content-length header
+        """
+        parsed = urlparse(spec)
+        if not parsed.scheme.endswith("+meca"):
+            return None
+        parsed = parsed._replace(scheme=parsed.scheme[:-5])
+        url = urlunparse(parsed)
+
+        r = self.session.head(url)
+        changes_with_content = r.headers.get("ETag") or r.headers.get("Content-Length")
```
Comment (sgaist) [inline repo2docker/contentproviders/meca.py:94]:
```suggestion
        headers = self.session.head(url).headers
        changes_with_content = headers.get("ETag") or headers.get("Content-Length")
```

File: repo2docker/contentproviders/meca.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+from hashlib import md5
+from os import path
+import tempfile
+import shutil
+import xml.etree.ElementTree as ET
+from zipfile import ZipFile, is_zipfile
+from urllib.parse import urlparse, urlunparse
+
+def get_hashed_slug(url, changes_with_content):
+    """Return a unique slug that is invariant to query parameters in the url"""
```
Comment (sgaist) [inline repo2docker/contentproviders/meca.py:13]:
```suggestion
    """Returns a unique slug that is invariant to query parameters in the url"""
```

**Review by sgaist — CHANGES_REQUESTED** [review]
The implementation functionally wise looks fine.

There are some minor suggestions.

I would also recommend to run `pre-commit`, it will cleanup the imports as well as some whites pace related things.

**Comment by stevejpurves** [comment]
thanks @sgaist 
I'm continuing to test locally this week and hope to pop this PR out of "draft" be the week end. I'm hitting some edge cases when:

- [] url specs contain multiple query params, these can be stripped after the first parameter
- [] the provider works when used in conjunction with binderhub but errors out when used from the commandline

**Comment by yuvipanda** [comment]
Hi @stevejpurves :) I just merged another new provider (CKAN), so wanted to check in to see if there's anything I can do to help this move forward!

**Comment by stevejpurves** [comment]
I've brought this up to date and also made some changes to allow a server to be launched from a local something.meca.zip file. Basically using a naming convention as way to initially detect a meca bundle specifically.

I guess that could be changed to check *any* local file for the expected contents, before returning true, but maybe this is fine for now.

I still intend to resolve:

- [ ] url specs contain multiple query params, these can be stripped after the first parameter
- [ ] the provider works when used in conjunction with binderhub but errors out when used from the command line
